patches/tModLoader/Terraria/ModLoader/Config/ConfigManager.cs

```diff
@@ -72,6 +74,113 @@ internal static void Add(ModConfig config)
 		configList2.Add(GeneratePopulatedClone(config));
 	}
 
+	internal static void FinishSetup()
+	{
+		// Register localization for all fields and properties that should show
+		foreach (var activeConfigs in ConfigManager.Configs) {
+			foreach (var config in activeConfigs.Value) {
+				try {
+					string modConfigLabelKey = GetModConfigLabelKey(config, throwErrors: true);
+					Language.GetOrRegister(modConfigLabelKey, () => Regex.Replace(config.Name, "([A-Z])", " $1").Trim());
+				}
+				catch (ValueNotTranslationKeyException e) {
+					e.additional = $"The ModConfig class '{config.Name}' caused this exception.";
+					e.Data["mod"] = config.Mod.Name;
+					throw;
+				}
+
+				RegisterLocalizationKeysForMembers(config, config.GetType());
+			}
+		}
+	}
+
+	private static void RegisterLocalizationKeysForMembers(ModConfig config, Type type)
+	{
+		foreach (PropertyFieldWrapper variable in ConfigManager.GetFieldsAndProperties(type)) {
+#pragma warning disable CS0618
+			if (Attribute.IsDefined(variable.MemberInfo, typeof(JsonIgnoreAttribute)) && !(Attribute.IsDefined(variable.MemberInfo, typeof(LabelAttribute)) || Attribute.IsDefined(variable.MemberInfo, typeof(ShowDespiteJsonIgnoreAttribute))))
+				continue;
+#pragma warning restore CS0618
+
+			try {
+				// TODO: Classes used in Generic types and Enums
+				// TODO: throw error after updating hjson somehow, so auto keys are populated?
+
+				// Register localization for classes added in this mod. This code handles the class itself, not the fields of the classes
+				if (variable.Type.IsClass && variable.Type.Assembly == type.Assembly) {
+#pragma warning disable CS0618 // Type or member is obsolete
+					var typeLabelObsolete = (LabelAttribute)Attribute.GetCustomAttribute(variable.Type, typeof(LabelAttribute));
+#pragma warning restore CS0618 // Type or member is obsolete
+
+					var typeLabel = (LabelKeyAttribute)Attribute.GetCustomAttribute(variable.Type, typeof(LabelKeyAttribute));
+					if (typeLabel != null) {
+						if (typeLabel.malformed)
+							throw new ValueNotTranslationKeyException($"{nameof(LabelKeyAttribute)} only accepts localization keys for the 'key' parameter.", errorOnType: true);
+					}
+					else { 
+						string typeLabelKey = config.Mod.GetLocalizationKey($"Configs.{variable.Type.Name}.Label");
+						string typeLabelValue = Regex.Replace(variable.Type.Name, "([A-Z])", " $1").Trim();
+						if (typeLabelObsolete != null)
+							typeLabelValue = typeLabelObsolete.LocalizationEntry;
+						Language.GetOrRegister(typeLabelKey, () => typeLabelValue);
+					}
+
+					var typeTooltip = (TooltipKeyAttribute)Attribute.GetCustomAttribute(variable.Type, typeof(TooltipKeyAttribute));
+					if (typeTooltip != null) {
+						if (typeTooltip.malformed)
```

Comment:
Fair bit of duplicate code here and elsewhere. Can `LabelKeyAttribute` and `TooltipKeyAttribute` share a common base class?